### PR TITLE
Add health check, functional tests, and CI

### LINKED ISSUE #89: CI testing
Maybe just a single github action that lanches the `example.graphml` graph and runs a few scenarios? To catch regressions and help discern what our supported feature set is!

### PR DESCRIPTION
**Pull request by pinheadmz**
See the successful CI run: https://github.com/pinheadmz/warnet/actions/runs/6383114707/job/17323124908

This PR gets us CI testing with three big steps, detailed below:

## Network health check RPCs

`info` reprints the `warnet.__str__()` output with this nice table:

```
--> warcli network info
Warnet:
        Temp Directory: /Users/matthewzipkin/.warnet/warnet/warnet
        Bitcoin Network: regtest
        Docker Network: warnet
        Subnet: 100.0.0.0/8
        Graph: DiGraph with 12 nodes and 22 edges
Tanks:
        Index   Version                  Conf                     Netem                    IPv4
        0       25.0                                              None                     100.229.34.56
        1       25.0                                              None                     100.194.17.152
        2       25.0                                              None                     100.48.139.54
        3       25.0                                              None                     100.148.146.231
        4       25.0                                              None                     100.6.80.159
        5       25.0                                              None                     100.149.173.199
        6       25.0                                              None                     100.22.254.194
        7       25.0                                              None                     100.207.152.71
        8       25.0                                              None                     100.175.150.96
        9       25.0                                              None                     100.204.160.179
        10      25.0                                              None                     100.231.24.178
        11      vasild/bitcoin#relay_tx_                          None                     100.7.209.52
```

`status` returns a JSON object with the docker container status for each tank. This RPC can be polled until all tanks are `"running"`. The CLI decodes the JSON response like this:

```
--> warcli network status
warnet_tank_000000: running
warnet_tank_000001: running
warnet_tank_000002: running
warnet_tank_000003: running
warnet_tank_000004: running
warnet_tank_000005: running
warnet_tank_000006: running
warnet_tank_000007: running
warnet_tank_000008: running
warnet_tank_000009: running
warnet_tank_000010: running
warnet_tank_000011: running
```

## Functional tests
All the goodies are in `test/`. There are a few test files that can be run individually during development, example:
```
$ python test/scenarios_test.py
```
Each individual test creates a `TestBase` instance which runs a warnet server in a subprocess and can create a network with a unique name and temporary directory. The output of the warnet server is captured to a file in another temporary location and that file can be read from continuously during the test. This allows us to monitor the server even though its in the background. That's especially nice during long waiting periods like while `docker-compose up` is running. `TestBase` also attempts to clean up if the test fails, the warnet server crashes, user hits ctrl-C, etc.

There is also a `data/` directory with a few example graphml files used in the tests. Hopefully writing new tests is self-explanatory based on the three examples I wrote so far.

## CI

Pretty simple, a github action that installs dependencies and then runs the individual tests. For now each test is run as a command in the ci workflow itself `ci.yml` so if you add a new test you need to be sure it is executed from that file. In the future we can work in a "test runner" like Bitcoin Core that runs a list of test files.

### CODE REVIEW

**Comment by pinheadmz** [comment]
Now running tests in parallel on ci: https://github.com/pinheadmz/warnet/actions/runs/6387282902/job/17335380750

File: test/data/v25_x_12.graphml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?><graphml xmlns="http://graphml.graphdrawing.org/xmlns">
+  <key attr.name="version" attr.type="string" for="node" id="version"/>
+  <key attr.name="bitcoin_config" attr.type="string" for="node" id="bitcoin_config"/>
+  <key attr.name="tc_netem" attr.type="string" for="node" id="tc_netem"/>
+  <graph edgedefault="directed">
+    <node id="0">
```
Comment (willcl-ark) [inline test/data/v25_x_12.graphml:6]: Why don't we just use say 4 nodes for this? Will be faster at testing everything?

They can each connect to 1 or 2 other nodes, and have space left for 1 or 2 organic outbounds

File: src/warnet/server.py [hunk 2]
```diff
@@ -291,6 +293,28 @@ def down(self, network: str = "warnet") -> str:
...
+
+
+    def status(self, network: str = "warnet") -> List[dict]:
+        """
+        Get running status of a warnet network named <network>
+        """
+        wn = Warnet.from_network(network)
+        stats = []
+        for tank in wn.tanks:
+            status = tank.container.status if tank.container is not None else None
```
Comment (willcl-ark) [inline src/warnet/server.py:311]: I kind of feel inclined here to use docker api to compare to our list of tanks, but it touches a lot of the other code and isn't worth the complexity (and yo've already got all this working)

**Review by willcl-ark — APPROVED** [review]
general ACK from me.

Left some comments, but I think we can move ahead with this and improve later?

**Comment by willcl-ark** [comment]
e.g. you can filter the docker api to get results like:

```
 {'Id': '1c779aaac627ed3d331435c0124b0845dd41ec8bf43bcf11cb4f9c54ddc840ab',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000002',
  'Status': 'running'},
 {'Id': '1e07c997d3b800a6df2e7874dbfe400434ff2e9c3291da825bc49339718fed5a',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000000',
  'Status': 'running'},
 {'Id': '7a1fa45191b2b3adfdf0e81e5349f72fd0c1730f972d80847c5566b9c32d85da',
  'Image': 'warnet_test_kl0lm1tn_dnsseed',
  'Name': 'warnet_test_kl0lm1tn_dns-seed',
  'Status': 'running'},
 {'Id': '6edfd9bf0eac34806cd3659119363c93f9a34e9a18f6814e0ccfab20476749c7',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000007',
  'Status': 'running'},
 {'Id': 'd4fd88e1c47a3163d818a35e13ef95dcd0eb56ec8a900420a197a1b21edfc286',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000008',
  'Status': 'running'},
 {'Id': '1d0472cf353fdde6ca944287f4ebd412ed0e3a4bdcdacf2f57b4515d94ed3a8d',
  'Image': 'prom/node-exporter:latest',
  'Name': 'warnet_test_kl0lm1tn_node-exporter',
  'Status': 'running'},
 {'Id': '7f4e6c2feb2594460739de8671a360a2e742df7d9c77aa788e67b10773cdd243',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000004',
  'Status': 'running'},
 {'Id': '12f5590fc72d66516326df08bfab5db651be65493391fa7a8922eb71e36037ec',
  'Image': 'jvstein/bitcoin-prometheus-exporter',
  'Name': 'warnet_test_kl0lm1tn_prometheus_exporter_000001',
  'Status': 'running'},
 {'Id': 'e691a722006ef7daff6e0622af085f8492c24328cff2fac2feed6092cf699588',
  'Image': 'warnet_test_kl0lm1tn_warnet_test_kl0lm1tn_tank_000000',
  'Name': 'warnet_test_kl0lm1tn_tank_000000',
  'Status': 'running'},
 {'Id': 'ac944b9f90e57931de356ee29c60f4536850bf65e9679b1c4dd2bab89f14759f',
  'Image': 'warnet_test_kl0lm1tn_warnet_test_kl0lm1tn_tank_000009',
  'Name': 'warnet_test_kl0lm1tn_tank_000009',
  'Status': 'running'}]
```

But I think what you have is pretty nice and suitable too

File: test/data/v25_x_12.graphml (hunk 1 above)
Comment (pinheadmz) [reply test/data/v25_x_12.graphml:6]: IMO 11 is kinda the minimum. The reason is bc core likes to establish 10 total connections (8 full + 2 blocks only) and if can't fill those 10 we get that endless log-spammy addrman message about trying new hosts. I'd say lets leave this for now and if you want to add tests that start with a smaller graph or whatever, ill review ;-)

File: src/warnet/server.py (hunk 2 above)
Comment (pinheadmz) [reply src/warnet/server.py:311]: `container` here is retrieved from the API


https://github.com/bitcoin-dev-project/warnet/blob/080156dbd74fb85c2ca181227edf9a12bf5406a1/src/warnet/tank.py#L137-L144

**Pull request merged**